### ISSUE TITLE
Required enforced even when not asked

### ISSUE THREAD

**Issue by lafont-e** [Jumpscale/go-raml#453]
in the github.com/zero-os/0-metadata/ project we have discovered that the field posix in the dir type on the github.com/zero-os/0-metadata/api_spec/main.raml is marked as "required: false" but the code generated complains when such field is not available.

In github.com/zero-os/0-metadata/tests/test_dir.py you can see that the test fails when the posix field is deleted in the creation of the dir object.

It fails the json validation even when not required.

**Comment by iwanbk** [comment]
> In github.com/zero-os/0-metadata/tests/test_dir.py you can see that the test fails when the posix field is deleted in the creation of the dir object.

I can't see the file at https://github.com/zero-os/0-metadata/tree/master/test

do you mean this?
https://github.com/zero-os/0-metadata/blob/pydevel2/test/test_dir.py

I'm not sure what happens with your test code. 

But i did a quick check for this `required: false`.
It works as expected.

raml
```yaml
#%RAML 1.0
title: goraml dir
description: goraml developers directory
baseUri: http://localhost:5000
mediaType: application/json

types:
  User:
    properties:
        username:
          type: string
        name:
          type: string
          required: false
```

generate the code
```
go-raml client --ramlfile ../../simple.raml --dir goramldir -l python --python-unmarshall-response=true
```

test

`name` is not required, shouldn't return error
```
$python3
Python 3.5.1 (default, Dec 27 2015, 02:23:23) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import goramldir.User as User
>>> user = User(username="iwan")
```

`username` is required, should return error
```python
>>> user = User(name="iwan")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ibk/pathgo/src/github.com/Jumpscale/go-raml/docs/tutorial/python/client/goramldir/User.py", line 37, in __init__
    self.username = client_support.set_property('username', data, data_types, False, [], False, True, class_name)
  File "/Users/ibk/pathgo/src/github.com/Jumpscale/go-raml/docs/tutorial/python/client/goramldir/client_support.py", line 132, in set_property
    raise ValueError(required_error.format(cls=class_name, prop=name))
ValueError: User: missing required property username
```

**Comment by zaibon** [comment]
The problem is not about the creation of the pythons objects. But in the validation in the handlers. In your example if we create the user with name = None the validation fails cause the type of name is none and not string.

**Comment by iwanbk** [comment]
> The problem is not about the creation of the pythons objects. But in the validation in the handlers. In your example if we create the user with name = None the validation fails cause the type of name is none and not string.

I've tested the handler, it was fine with
```
{
    "username": "iwan"
}
```

But failed with 
```
{
    "name": null,
    "username": "iwan"
}
```

Which was produced by the class.

So, the problem is not in the handler, but in the class' `__str__` method.
Which implemented in `client_support.py`

```python
$python3
Python 3.5.1 (default, Dec 27 2015, 02:23:23) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import goramldir.User as User
>>> user = User(username="iwan")                                                         
>>> print(user)
{
    "name": null,
    "username": "iwan"
}
```

**Comment by iwanbk** [comment]
This patch seems to fix this issue, but need to check it more.
I also really wonder why we didn't get this bug earlier.

```diff
diff --git a/codegen/templates/python/client_support.tmpl b/codegen/templates/python/client_support.tmpl
index 8bcfecf..6b0361f 100644
--- a/codegen/templates/python/client_support.tmpl
+++ b/codegen/templates/python/client_support.tmpl
@@ -166,7 +166,7 @@ def to_dict(cls, convert_datetime=True):
             return [todict(v) for v in obj]
         elif hasattr(obj, "__dict__"):
             for key, value in obj.__dict__.items():
-                if not callable(value) and not key.startswith('_'):
+                if value and not callable(value) and not key.startswith('_'):
                     data[key] = todict(value)
             return data
         else:
```

**Comment by iwanbk** [comment]
> This patch seems to fix this issue, but need to check it more.

@zaibon @lafont-e would be great if you guys could help me to check it.

To make things easier, you don't need to modify the template, you could modify the generated `client.py` instead.

**Comment by iwanbk** [comment]
thinking of it again, not sure, `null` might be valid
```
{
    "name": null,
    "username": "iwan"
}
```
So this json might be valid.

**Comment by iwanbk** [comment]
> I also really wonder why we didn't get this bug earlier.

#455 should prevent this

**Comment by iwanbk** [comment]
Fixed at #456, now working on #455 with this issue as the first test case

**Comment by iwanbk** [comment]
> Fixed at #456, now working on #455 with this issue as the first test case

Some portion of #455 which relates to this issue has been added to #456 PR.

@zaibon @lafont-e please check
